File: lexy-app/backend/routers/word_lists.py
from fastapi import APIRouter, Depends, HTTPException, Path
from fastapi.responses import PlainTextResponse

from ..core.deps import get_current_user
from ..database import get_pool
from ..models.schemas import (
    WordListCreate,
    WordListDetail,
    WordListMarkLearningResult,
    WordListSummary,
)
from ..services import word_list_service

router = APIRouter(prefix="/word-lists", tags=["word-lists"])

# Every route is scoped by (list_id, user_id) in the service layer and answers
# 404 — not 403 — when the row belongs to someone else. A 403 would confirm the
# id exists, letting one user enumerate another's lists.
_NOT_FOUND = "Word list not found"


@router.post("", response_model=WordListDetail, status_code=201)
async def create_word_list(
    body: WordListCreate,
    pool=Depends(get_pool),
    current_user: dict = Depends(get_current_user),
):
    """
    Create a list from pasted or uploaded words.

    Surfaces are stripped, deduped case-insensitively (first spelling wins),
    and resolved against word_table language-scoped + case-insensitively.
    Nothing is dropped: a surface with no match is stored as `unresolved`, and
    one with several matches as `ambiguous` rather than being bound to an
    arbitrary sense.
    """
    try:
        return await word_list_service.create_list(
            pool,
            str(current_user["user_id"]),
            name=body.name,
            language=body.language,
            words=body.words,
            description=body.description,
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("", response_model=list[WordListSummary])
async def get_word_lists(
    pool=Depends(get_pool),
    current_user: dict = Depends(get_current_user),
):
    return await word_list_service.list_lists(pool, str(current_user["user_id"]))


@router.get("/{list_id}", response_model=WordListDetail)
async def get_word_list(
    list_id: int = Path(..., ge=1),
    pool=Depends(get_pool),
    current_user: dict = Depends(get_current_user),
):
    result = await word_list_service.get_list(
        pool, str(current_user["user_id"]), list_id,
    )
    if result is None:
        raise HTTPException(status_code=404, detail=_NOT_FOUND)
    return result


@router.get("/{list_id}/export", response_class=PlainTextResponse)
async def export_word_list(
    list_id: int = Path(..., ge=1),
    pool=Depends(get_pool),
    current_user: dict = Depends(get_current_user),
):
    """One surface per line, insertion order, unresolved and ambiguous included."""
    text = await word_list_service.export_list(
        pool, str(current_user["user_id"]), list_id,
    )
    if text is None:
        raise HTTPException(status_code=404, detail=_NOT_FOUND)
    return PlainTextResponse(
        content=text,
        media_type="text/plain",
        headers={"Content-Disposition": f'attachment; filename="word-list-{list_id}.txt"'},
    )


@router.post("/{list_id}/mark-unknown-learning", response_model=WordListMarkLearningResult)
async def mark_unknown_learning(
    list_id: int = Path(..., ge=1),
    pool=Depends(get_pool),
    current_user: dict = Depends(get_current_user),
):
    """
    Mark every resolved-but-unknown entry as 'learning'.

    Runs through progression_service with `status_marked_learning` +
    `status_override="learning"` — the same path as the words.py status route —
    so levels, auto-promotion and both SRS cards match a hand-marked word.
    Unresolved and ambiguous entries are skipped; already learning/known
    entries are untouched, so repeat calls are idempotent.
    """
    result = await word_list_service.mark_unknown_as_learning(
        pool, str(current_user["user_id"]), list_id,
    )
    if result is None:
        raise HTTPException(status_code=404, detail=_NOT_FOUND)
    return result


@router.delete("/{list_id}", status_code=204)
async def delete_word_list(
    list_id: int = Path(..., ge=1),
    pool=Depends(get_pool),
    current_user: dict = Depends(get_current_user),
):
    deleted = await word_list_service.delete_list(
        pool, str(current_user["user_id"]), list_id,
    )
    if not deleted:
        raise HTTPException(status_code=404, detail=_NOT_FOUND)

File: lexy-app/frontend/src/api/wordLists.ts
import { apiUrl } from './_baseUrl';
import { assertOk, assertOkJson } from './_http';

/**
 * Vocabulary lists — paste/upload a word list, see what you already know.
 *
 * Five entry states, mirroring backend/services/word_list_service.py:
 *   known | learning | unknown  — bound to a catalog word
 *   unresolved                  — no match in the dictionary for that language
 *   ambiguous                   — several matches; the backend deliberately
 *                                 refuses to guess which sense was meant
 */
export type WordListEntryStatus =
    | 'known'
    | 'learning'
    | 'unknown'
    | 'unresolved'
    | 'ambiguous';

/**
 * Which catalog a resolved entry is bound to. Set even when `item_id` is null —
 * it records which table the surface was looked up in, so an unresolved entry
 * still says what it was trying to be.
 */
export type WordListItemType = 'word' | 'phrase';

export interface WordListEntry {
    id: number;
    surface: string;
    item_id: number | null;
    item_type: WordListItemType;
    status: WordListEntryStatus;
}

export interface WordListSummary {
    list_id: number;
    name: string;
    language: string;
    description: string | null;
    created_at: string;
    total: number;
    /**
     * Built-in list, shared by every user and not user-editable (backend
     * migration 037). The backend refuses to delete one and suppresses its own
     * writes to shared rows, so hiding destructive controls in the UI is a
     * courtesy, not the guarantee — the guarantee is server-side.
     *
     * Optional so a response from a backend older than 037 still parses.
     */
    is_system?: boolean;
}

export interface WordListDetail extends Omit<WordListSummary, 'total'> {
    total: number;
    counts: Record<WordListEntryStatus, number>;
    entries: WordListEntry[];
}

export interface MarkLearningResult {
    list_id: number;
    marked: number;
    marked_item_ids: number[];
    skipped_unresolved: number;
    skipped_ambiguous: number;
    /**
     * Eligible entries this call did not reach — marking is capped at
     * MAX_LIST_WORDS per request so one click on a 5,000-item built-in list
     * cannot flood a user's reviews. The call is idempotent, so the fix is
     * simply to call again. Optional for a pre-cap backend.
     */
    remaining?: number;
    /** True while `remaining > 0`. */
    capped?: boolean;
}

/** Max words per list — must match WORD_LIST_MAX_WORDS in the backend schemas. */
export const MAX_LIST_WORDS = 500;

/**
 * Split pasted text or an uploaded .txt into words.
 *
 * Splits on newlines, commas, tabs and semicolons so the same parser handles a
 * one-per-line list and a comma-separated one. Spaces are NOT separators —
 * multi-word surfaces stay intact for the phrase catalog. Blank entries are
 * dropped here so the count shown to the user matches what gets posted; the
 * backend dedupes case-insensitively.
 */
export function parseWordInput(raw: string): string[] {
    return raw
        .split(/[\n,;\t]+/)
        .map(w => w.trim())
        .filter(Boolean);
}

function authHeaders(token: string): HeadersInit {
    return { Authorization: `Bearer ${token}` };
}

export async function createWordList(
    token: string,
    name: string,
    language: string,
    words: string[],
): Promise<WordListDetail> {
    const res = await fetch(apiUrl('/api/v1/word-lists'), {
        method: 'POST',
        headers: { ...authHeaders(token), 'Content-Type': 'application/json' },
        body: JSON.stringify({ name, language, words }),
    });
    return assertOkJson<WordListDetail>(res, 'Failed to create word list');
}

export async function listWordLists(token: string): Promise<WordListSummary[]> {
    const res = await fetch(apiUrl('/api/v1/word-lists'), { headers: authHeaders(token) });
    return assertOkJson<WordListSummary[]>(res, 'Failed to load word lists');
}

export async function getWordList(token: string, listId: number): Promise<WordListDetail> {
    const res = await fetch(apiUrl(`/api/v1/word-lists/${listId}`), {
        headers: authHeaders(token),
    });
    return assertOkJson<WordListDetail>(res, 'Failed to load word list');
}

/** The list as plain text, one surface per line — unresolved/ambiguous included. */
export async function exportWordList(token: string, listId: number): Promise<string> {
    const res = await fetch(apiUrl(`/api/v1/word-lists/${listId}/export`), {
        headers: authHeaders(token),
    });
    // Not assertOkJson: the endpoint returns text/plain, so parsing as JSON
    // would throw on the success path.
    await assertOk(res, 'Failed to export word list');
    return res.text();
}

export async function markUnknownAsLearning(
    token: string,
    listId: number,
): Promise<MarkLearningResult> {
    const res = await fetch(apiUrl(`/api/v1/word-lists/${listId}/mark-unknown-learning`), {
        method: 'POST',
        headers: authHeaders(token),
    });
    return assertOkJson<MarkLearningResult>(res, 'Failed to mark words as learning');
}

export async function deleteWordList(token: string, listId: number): Promise<void> {
    const res = await fetch(apiUrl(`/api/v1/word-lists/${listId}`), {
        method: 'DELETE',
        headers: authHeaders(token),
    });
    await assertOk(res, 'Failed to delete word list');
}
